NPE when running Zipkin in CF

when running Zipkin in CF I got such an interesting exception

```
 2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] Caused by: java.lang.NullPointerException: null
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.CorrectForClockSkew.adjustTimestamps(CorrectForClockSkew.java:182) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.CorrectForClockSkew.adjust(CorrectForClockSkew.java:112) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.CorrectForClockSkew.adjust(CorrectForClockSkew.java:128) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.CorrectForClockSkew.apply(CorrectForClockSkew.java:97) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.CorrectForClockSkew.apply(CorrectForClockSkew.java:51) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.V2SpanStoreAdapter$GetTracesMapper.map(V2SpanStoreAdapter.java:173) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.V2SpanStoreAdapter$GetTracesMapper.map(V2SpanStoreAdapter.java:165) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin2.Call$Mapping.doExecute(Call.java:237) ~[io.zipkin.zipkin2-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin2.Call$Base.execute(Call.java:379) ~[io.zipkin.zipkin2-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.internal.V2SpanStoreAdapter.getTraces(V2SpanStoreAdapter.java:44) ~[io.zipkin.java-zipkin-2.7.1.jar!/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at zipkin.server.internal.ZipkinQueryApiV1.getTraces(ZipkinQueryApiV1.java:106) ~[classes/:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at sun.reflect.GeneratedMethodAccessor60.invoke(Unknown Source) ~[na:na]
2018-04-19T16:29:52.743+02:00 [APP/PROC/WEB/0] [OUT] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_162]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_162]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133) ~[spring-web-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:827) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:738) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967) ~[spring-webmvc-4.3.14.RELEASE.jar!/:4.3.14.RELEASE]
2018-04-19T16:29:52.744+02:00 [APP/PROC/WEB/0] [OUT] at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901) ~[spring-webmvc-4.3.14.RELEA
```

https://github.com/openzipkin/zipkin/blob/master/zipkin/src/main/java/zipkin/internal/CorrectForClockSkew.java#L182

```java
  // Search for a local span on the skewed endpoint
    for (int i = 0, length = span.binaryAnnotations.size(); i < length; i++) {
      BinaryAnnotation b = span.binaryAnnotations.get(i);
      if (b.endpoint == null) continue;
      if (b.key.equals(Constants.LOCAL_COMPONENT) && ipsMatch(skew.endpoint, b.endpoint)) {
        return span.toBuilder().timestamp(span.timestamp - skew.skew).build();
      }
    }
    return span;
```

a couple of NPEs possible here 

`return span.toBuilder().timestamp(span.timestamp - skew.skew).build();`

--- a/zipkin/src/main/java/zipkin/internal/CorrectForClockSkew.java
+++ b/zipkin/src/main/java/zipkin/internal/CorrectForClockSkew.java
@@ -155,13 +155,13 @@ public final class CorrectForClockSkew {
   /** If any annotation has an IP with skew associated, adjust accordingly. */
   static Span adjustTimestamps(Span span, ClockSkew skew) {
     List<Annotation> annotations = null;
-    Long annotationTimestamp = null;
+    Long spanTimestamp = span.timestamp, annotationTimestamp = null;
     for (int i = 0, length = span.annotations.size(); i < length; i++) {
       Annotation a = span.annotations.get(i);
       if (a.endpoint == null) continue;
       if (ipsMatch(skew.endpoint, a.endpoint)) {
         if (annotations == null) annotations = new ArrayList<>(span.annotations);
-        if (span.timestamp != null && a.timestamp == span.timestamp) {
+        if (spanTimestamp != null && a.timestamp == spanTimestamp) {
           annotationTimestamp = a.timestamp;
         }
         annotations.set(i, a.toBuilder().timestamp(a.timestamp - skew.skew).build());
@@ -175,11 +175,12 @@ public final class CorrectForClockSkew {
       return builder.build();
     }
     // Search for a local span on the skewed endpoint
+    if (spanTimestamp == null) return span; // We can't adjust something lacking a timestamp
     for (int i = 0, length = span.binaryAnnotations.size(); i < length; i++) {
       BinaryAnnotation b = span.binaryAnnotations.get(i);
       if (b.endpoint == null) continue;
       if (b.key.equals(Constants.LOCAL_COMPONENT) && ipsMatch(skew.endpoint, b.endpoint)) {
-        return span.toBuilder().timestamp(span.timestamp - skew.skew).build();
+        return span.toBuilder().timestamp(spanTimestamp - skew.skew).build();
       }
     }
     return span;

--- a/zipkin/src/test/java/zipkin/internal/CorrectForClockSkewTest.java
+++ b/zipkin/src/test/java/zipkin/internal/CorrectForClockSkewTest.java
@@ -242,8 +242,13 @@ public class CorrectForClockSkewTest {
     Span rpcSpan = childSpan(rootSpan, APP_ENDPOINT, now + networkLatency, 1000L, skew);
     Span local = localSpan(rpcSpan, APP_ENDPOINT, rpcSpan.timestamp + 5, 200L);
     Span local2 = localSpan(local, APP_ENDPOINT, local.timestamp + 10, 100L);
+    Span local3 = Span.builder().traceId(local2.traceId).parentId(local2.id).id(local2.id +1)
+      .name("crappylc") // missing timestamp duration and even missing endpoint
+      .addBinaryAnnotation(BinaryAnnotation.create(LOCAL_COMPONENT, "lc",null))
+      .build();
 
-    List<Span> adjustedSpans = CorrectForClockSkew.apply(asList(rpcSpan, rootSpan, local, local2));
+    List<Span> adjustedSpans =
+      CorrectForClockSkew.apply(asList(rpcSpan, rootSpan, local, local2, local3));
 
     Span adjustedLocal = getById(adjustedSpans, local.id);
     assertThat(local.timestamp - skew)
@@ -252,6 +257,10 @@ public class CorrectForClockSkewTest {
     Span adjustedLocal2 = getById(adjustedSpans, local2.id);
     assertThat(local2.timestamp - skew)
         .isEqualTo(adjustedLocal2.timestamp.longValue());
+
+    Span adjustedLocal3 = getById(adjustedSpans, local3.id);
+    assertThat(adjustedLocal3)
+      .isEqualTo(local3); // no change
   }
 
   @Test

